Realm/RLMQueryUtil.mm

```diff
@@ -625,6 +635,50 @@ void add_collection_operation_constraint(NSPredicateOperatorType operatorType,
     }
 }
 
+// Diacritic-sensitive version.
+void QueryBuilder::add_substring_constraint(const StringData& value, Query&& condition) {
+    // Foundation always returns false for substring operations with a RHS of null or "".
+    m_query.and_query(value.size()
+                      ? std::move(condition)
+                      : std::unique_ptr<Expression>(new FalseExpression));
+}
+
+// Diacritic-sensitive version.
+void QueryBuilder::add_substring_constraint(const Columns<String>& value, Query&& condition) {
+    // Foundation always returns false for substring operations with a RHS of null or "".
+    // We don't need to concern ourselves with the possibility of column2 traversing a link list
+    // and producing multiple values per row as such expressions will have been rejected.
+    m_query.group();
+    m_query.and_query(value != null() && value != "");
+    m_query.and_query(std::move(condition));
+    m_query.end_group();
+}
+
+// Diacritic-insensitive version.
+template<typename ConstraintAdder, typename P>
+void QueryBuilder::add_substring_constraint(const StringData& value,
+                                            const P& comparator,
+                                            ConstraintAdder&& add_constraint) {
+    if (value.size()) {
+        add_constraint(comparator);
```

Comment:
It seems surprising that this function takes `add_constraint` and expects it to mutate `m_query`. It'd make more sense for this to take the string comparison query as an argument and call `and_query` here to combine them instead.

I think that would also allow these functions to be merged with the diacritic-sensitive versions?